[LIBCLOUD-1022] Enhance digitalocean dns driver with `ttl` parameter

kiniou opened the pull request:
## Enhance digitalocean dns driver with `ttl` parameter

### Description

DigitalOcean provides a parameter to update dns records with `ttl` but the current driver's code actually doesn't support it.

For reference:

- https://developers.digitalocean.com/documentation/v2/#update-a-domain-record
- https://developers.digitalocean.com/documentation/v2/#create-a-new-domain-record

### Status

ready for review


### Checklist (tick everything that applies)

- [X] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [X] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Add cinder support LIBCLOUD-874

micafer committed:
Move sec groups to neutron api LIBCLOUD-874

micafer committed:
Move sec groups to neutron api LIBCLOUD-874

micafer committed:
Move floatingips to neutron api LIBCLOUD-874

micafer committed:
Fix _to_port to make it work with old versions

micafer committed:
Remove print

micafer committed:
Update comment

micafer committed:
add direction in security_group_rule creation

Commit:
Added new code for xml parsing to common/nttcis

micafer committed:
Add create and delete subnets

micafer committed:
Remove prints

micafer committed:
Solve PR issues

micafer committed:
Fix conn issues

Commit:
added file test_lib_list.py

Commit:
Merge branch 'drs' into feature_drs hopefully get all changes while preserving the drs and tests directories

Commit:
Added abstract Driver class in drs/drivers/nttcis

Commit:
the drs driver does function

Commit:
merged with drs

Commit:
removed dd prefix from region on tests files

Commit:
changed _to_nodes and _to_node and drivers/nttcis.py to handle node retrieval by process_xml

Commit:
Test failed DRS creation not supported between the selected data centers

Commit:
moved create_consistency_group to drs/drivers/nttcis.py from compute/drivers/nttcis.py. Started testing delete

Commit:
Merge branch 'feature_drs' of https://bitbucket.mcp-services.net/scm/devops/libcloud into dynamic

Commit:
can list cg groups

Commit:
added directories and started files for drs/test suite

Commit:
added fixtures for drs ineligible error and listing cg's and completed tests

Commit:
commented out several classes in common/nttcis module, added tests for get single cg

Commit:
driver: update Exoscale info

Signed-off-by: Yoan Blanc <yoan.blanc@exoscale.ch>

Commit:
added listing snapshots, edited the XmlToDictConfig class to handle XML from snapshots, added integrations test, and a fixture for unit tests

Commit:
added more tests to tests and test directory.  nttcis module for drs uses a decorator to list/filter consistency groups

Commit:
Added documentation for drs via sphinx and added tests for expand journal.

Commit:
added auto generated rst files to ../docs/drs/

Commit:
added addtional tests and the functionality to choose snapshots by filtering create times

Commit:
drs finished...need to run tox, pull latest and pull request

kiniou committed:
add `ttl` parameters to digitalocean record operations

kiniou commented:
Hi 👋

I was not sure if the `ttl` parameters should go directly in methods signature or in `extra` arguments.

There seems to be some inconsistence between some drivers (eg. [powerdns](https://github.com/apache/libcloud/blob/trunk/libcloud/dns/drivers/powerdns.py#L189)) and [documentation](https://libcloud.readthedocs.io/en/latest/dns/examples.html#create-a-record-with-a-custom-ttl).

Just let me know if i need to do the switch 😉.

Commit:
fixed drs examples in docs/examples/drs/nttcis; added imports for MutableSequence and Mapping based on python vesions in libcloud/common/nttcis;

added code blocks for dynamic class creation from xml as opposed to pick and choose properties from xml in common/nttcis; firewall rules are dynamic

Commit:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into feature_drs

Commit:
renamed file after testing .pylintrc.bak -> .pylintrc

Commit:
removed indent from docstring of NttCisDRSDriver

Commit:
removed more indents on libcloud/drs/drivers/nttcis.py

Commit:
edited index.rst under drs/drivers in docs

Commit:
deleted uneeded files

Commit:
Started ssl import and offload

Commit:
Finished ssl import and list

Commit:
added ssl cert chain imports, listing cert chains, get a cert chain.  Created fixtures and pytest tests for certs

Commit:
added ssl offload profiles as well as listing and getting profiles

ktdreyer committed:
container/utils: improve docstring for RegistryClient

Prior to this change, the docstring for the RegistryClient class was
copy-and-pasted from the HubClient class.

Add the definition for the "host" parameter, and change the descriptions
to reflect the generalized nature of this RegistryClient class.

ktdreyer committed:
docs: autodoc libcloud.container.utils.docker.RegistryClient

Add libcloud.container.utils.docker.RegistryClient to the documentation
that Sphinx generates.

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

Commit:
finished tests and tox

Commit:
LIBCLOUD-1024 commit for pull request

Commit:
removed indentation of list in in list_consisitency_groups

Commit:
added pyopenssl to travis-coverage in tox.init and edited nttcis.rst in drs/drivers

vdloo committed:
Add changes for #1254

Closes #1254

MikeAT committed:
[LIBCLOUD-1025] Add a new DNS driver for RcodeZero DNS.

MikeAT committed:
[LIBCLOUD-1025] fixed travis errors.

vdloo committed:
implement OpenStack_1_1_NodeDriver ex_get_snapshot

to list a volume snapshot by ID, similar to `ex_get_volume`

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

Commit:
added listing geographical regions

Commit:
removed changes made and will create new branch

tonybaloney commented:
thanks for the contribution @kiniou 
the `ttl` parameter should go in the method signature for zones, and in the `extra` dict for records. I appreciate that makes little sense!

tonybaloney commented on libcloud/dns/drivers/digitalocean.py:
```diff
@@ -178,6 +178,10 @@ def create_record(self, name, zone, type, data, extra=None):
             except KeyError:
                 params['weight'] = 'null'
+            try:
+                params['ttl'] = extra['ttl']
```
`params['ttl'] = extra.get('ttl', None)` is a little cleaner

kiniou replied:
:ok_hand: I also thought it would be cleaner to do `extra.get` but I thought the actual try-catch based code was following some coding rules i should follow. Maybe it would be more readable to do this on the others parameters. 

Anyway, I'll edit PR with your suggestion :wink:

kiniou replied:
According to the relevant DigitalOcean API documentations, the `ttl` can not be nulled and I've just tested the change to `extra.get('ttl', None)` and the API response fails with an error 500.

I've looked for a better syntactic way but i think the try-catch structure is the only way to not include this parameter if it does not exists.

vdloo replied:
you could do something like this:
```
if 'ttl' in extra:
    params['ttl'] = extra['ttl']
```

kiniou replied:
Yes ! :man_facepalming:
Thank you !

Commit:
Added more tests for mcp1.0 for greater coverage

vdloo committed:
Add changes for #1257

Closes #1257

iler committed:
[LIBCLOUD-1026] Update documentation for UpCloud driver

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

vdloo committed:
Add changes for #1259

Closes #1259

Commit:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into feature_ssl_offload

Commit:
removing drs componoents

micafer committed:
Change order in gets

micafer committed:
Complete docstring

micafer committed:
Improve comment

vdloo committed:
list calls paginate in OpenStackNodeDriver

the default max_limits for OpenStack is 1000. If you have more than 1000
resources (i.e. snapshots) then everything but the newest 1000 will not be
listed. If you set the max_limits lower even less will not be returned, etc.
This change implements pagination for ex_list_snapshots, ex_list_ports,
list_volumes and list_nodes in the OpenStack_2_NodeDriver.

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

ktdreyer committed:
openstack: handle missing user "enabled" attribute

Keystone may not always return an "enabled" attribute for a user
account. Prior to this change, we could crash if we queried a user that
lacked an "enabled" attribute. Gracefully handle this case by setting
"enabled" to None.

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

Commit:
added lots of server snapshot tests

Commit:
removed drs from provider feature matrix

Commit:
removed referenece to DRSDriver in feature_matrix

Commit:
removed reference to DRS in index.rst

Commit:
removed yet another reference to DRS in index.rst

micafer committed:
Fix OpenStackException issue

vdloo committed:
Add changes for #1254

Closes #1254

vdloo committed:
Add changes for #1261

Closes #1261

vdloo committed:
fix comment ProviderError dereived -> derived

Commit:
removed drs directories

micafer committed:
Fix error in ex_list_ports if created_at is not in result

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

ktdreyer committed:
tests: use assertIsNone

Python 2.7 introduced assertIsNone(). Change all our
"assertEqual(value, None)" statements to "assertIsNone(value)".

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

vdloo committed:
Add changes for #1264

Closes #1264

Commit:
moved much of drs functionality into the compute driver, last test coded far was stopping drs_preview_failover

Commit:
added tests for fail create and fail failover with invalid status

vdloo committed:
Add changes for #1241

Closes #1242

vdloo committed:
Add changes for #1263

Closes #1263

Commit:
completed tests for drs without a drs driver

Commit:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into feature_ssl_offload

Commit:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into feature_drs_without_driver

Commit:
Changed typo. Added pythoh-dateutil to tox.

Commit:
Removed requirements for python-dateutil for test/compute/test-nttcis.py

Commit:
made pyopenssl optional
Closes #1255

MikeAT committed:
fixed documentation

MikeAT committed:
use default-hostname as a class-attribute

MikeAT committed:
refactore double negotiation

MikeAT committed:
add method get_record

MikeAT committed:
regenerated supported methods/providers table

MikeAT committed:
fixed whitespaces

MikeAT committed:
fixed python3 issue and tests

vdloo committed:
fix OpenStack_2_NodeDriver.detach_volume

Since https://github.com/apache/libcloud/pull/1242 the
OpenStack_2_NodeDriver uses a volume object returned via
volumev2_connection. This object will not contain serverId
but server_id. This PR makes the detach_volume method
compatible with v2 volumes.

Fixes:
```
return connection.detach_volume(volume)
File "/usr/local/venv/project/src/apache-libcloud/libcloud/compute/drivers/openstack.py", line 328, in detach_volume
(attachment['serverId'], attachment['id']),
KeyError: 'serverId'
```

kiniou committed:
[LIBCLOUD-1022] check 'ttl' presence instead of try-catching

Commit:
merged latest trunk

Commit:
replaced Dimension Data with NTTC-CIS in some doc strings in loadbalancers/drivers/nttcis.py

Commit:
added an extra assertion to DRS tests

micafer committed:
Set suspended state

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

vdloo committed:
Add changes for #1269

Closes #1269

Commit:
Added 'ex' prefix to all DRS related methods

Commit:
Fixed lint issues

Commit:
fixed PEP 8 line too long 4855 & 4875

Commit:
fixed doc strings

Commit:
removeed doc string indentations

tonybaloney committed:
Merge branch 'libcloud_1268' into trunk
Closes #1268

tonybaloney committed:
Update docstrings for NTTC

Commit:
removed dimension data from nttcis.rst in docs/loadbalancers

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

vdloo committed:
fix nttc sphinx error

fixes:
```
Warning, treated as error:
/home/travis/build/apache/libcloud/libcloud/compute/drivers/nttcis.py:docstring of libcloud.compute.drivers.nttcis.NttCisNodeDriver.ex_delete_consistency_group:6:Unexpected indentation.
```
introduced in https://github.com/apache/libcloud/commit/6ae3a29ba5522c566b893f6994cc1748c1c7b149

vdloo committed:
Add changes for #1271

Closes #1271

vdloo committed:
Add changes for #1270

Closes #1270

MonsieurV committed:
Fix _is_gcs_s3 to accept access keys of more than 20 chars

Google Cloud Storage currently emits access keys of more than 20 characters

Signed-off-by: Rick van de Loo <rickvandeloo@gmail.com>

mitar committed:
Updating command line arguments to current version.

mitar committed:
Issues are not on GitHub.

vdloo committed:
Add changes for #1272

Closes #1272

Commit:
Adds missing docs for param ``ex_prefix`` within list_container_objects method.

Also adds ex_prefix functionality to the DummyStore, since all other stores have this feature.

tonybaloney committed:
Merge pull request #1256 from MikeAT/Libcloud-1025_add_RcodeZero_DNS_Driver

[LIBCLOUD-1025] Add a new DNS driver for RcodeZero DNS.

tonybaloney committed:
Update CHANGES.rst

Add changes for #1256

tonybaloney committed:
Merge pull request #1275 from richardARPANET/corrects-list-container-objects-docs

Adds missing docs for param ex_prefix & adds to DummyStore

tonybaloney committed:
Update CHANGES.rst

Changes for #1275

tonybaloney committed:
Merge pull request #1274 from mitar/patch-2

Issues are not on GitHub

tonybaloney committed:
Merge pull request #1273 from mitar/patch-1

Updating command line arguments to current version

tonybaloney committed:
Update CHANGES.rst

bmwiedemann committed:
Make tests pass after 2031

note: tests completed successfully on i586 and x86_64 in 2034

ryanpetrello committed:
Use cryptography (not PyCrypto) for GCE service account authentication

see: https://issues.apache.org/jira/browse/LIBCLOUD-1015

ryanpetrello committed:
replace all usage of PyCrypto with python-cryptography

ryanpetrello committed:
replace all usage of PyCrypto with python-cryptography

Kami committed:
Fix lint.

Kami committed:
Use self.assertEqual instead of assert.

Kami committed:
Pin cryptography to see if that helps with tests failing on travis.

Kami committed:
Add changelog entry.

Kami committed:
Merge branch 'gce-crypto' of https://github.com/ryanpetrello/libcloud into gce-crypto

Kami committed:
Merge pull request #1280 from ryanpetrello/gce-crypto

Replace all usage of PyCrypto with python-cryptography

Kami committed:
Merge branch 'FTBFS2031' of https://github.com/bmwiedemann/libcloud into FTBFS2031

Kami committed:
Make sure we always cast timestamp to string - this way it also works
with 64 bit ints.

Kami committed:
Merge pull request #1279 from bmwiedemann/FTBFS2031

Make tests pass after 2031

Commit:
LIBCLOUD-1040_CloudSigma_provider_ex_list_subscriptions_fails_if_your_subscription_includes_free_tier_resources

Kami committed:
Merge branch 'fix-detach-volume-OpenStack_2_NodeDriver' of https://github.com/vdloo/libcloud into vdloo-fix-detach-volume-OpenStack_2_NodeDriver

Resolves #1267

Kami committed:
Add changelog entry.

Kami committed:
Merge pull request #1251 from exoscale/fresher-exoscale

driver: update Exoscale info

Kami committed:
Update tox dependency on travis.

Kami committed:
Fix a bug with ECS tests using wrong hostname which caused tests to fail
on Travis with newer version of requests.

Kami committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into trunk

Kami committed:
Update various test dependencies (flake8, pylint, astroid, tox).

Kami committed:
Update pylint config so it works with new version of pylint.

Kami committed:
Add a workaround for false positives detected by new version of pylint.

Kami committed:
Fix various lint issues detected by new version of flake8

Kami committed:
Fix more false positives detected by new version of pylint.

Kami committed:
Fix more false positives.

Kami committed:
Upgrade astroid dependency.

Kami committed:
Add a workaround for Python 3.7 build failure.

It looks like the build fails on travis because of old setuptools.

Kami committed:
Merge pull request #1289 from Kami/fix_build

Add a workaround for Python 3.7 build failure on Travis

Kami committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/libcloud into update_test_dependencies
Try a workaround for failing py3.7 build.

Kami committed:
Merge pull request #1288 from Kami/update_test_dependencies

Update test dependencies

Kami committed:
Update CloudSigma driver to correctly handle subscription objects with
start_time and / or end_time attributes.

Kami committed:
Add changelog entry for it.

Kami committed:
Merge branch 'LIBCLOUD-1040_CloudSigma_provider_ex_list_subscriptions_fails_if_your_subscription_includes_free_tier_resources' of https://github.com/aki-k/libcloud into LIBCLOUD-1040_CloudSigma_provider_ex_list_subscriptions_fails_if_your_subscription_includes_free_tier_resources

Kami committed:
Merge pull request #1284 from aki-k/LIBCLOUD-1040_CloudSigma_provider_ex_list_subscriptions_fails_if_your_subscription_includes_free_tier_resources

LIBCLOUD-1040_CloudSigma_provider_ex_list_subscriptions_fails_if_your…

Kami committed:
Merge pull request #1262 from vdloo/dereived-derived-provider-error

fix comment ProviderError dereived -> derived

Kami commented:
Re ``ttl`` in the ``extra`` attribute - it looks like a lot of the existing drivers support TTL attribute for records so I think it would be time to "promote" this attribute from ``extra`` to ``ttl`` argument in create / update record methods in the next major release.

For backward compatibility reasons we should still support ``ttl`` in extra, with ``ttl` having precedence over extra attribute.

Kami committed:
Merge branch 'LIBCLOUD-1022_digitalocean-dns-record-ttl-parameter' of https://github.com/kiniou/libcloud into LIBCLOUD-1022_digitalocean-dns-record-ttl-parameter

Kami committed:
Don't explode if "ttl" attribute is not present, update affected tests.

Kami committed:
Add changelog entry.

Kami commented:
Sorry for the delay.

I made a small change (ec563f995c5080acba0087c20408a13df58b3d51) and merged changes into trunk. Thanks again for your contribution.
